pkg/pool-utils/contracts/rates/PriceRateCache.sol

```diff
@@ -34,60 +34,66 @@ library PriceRateCache {
+    function getCurrentRate(bytes32 cache) internal pure returns (uint256) {
+        return cache.decodeUint(_CURRENT_PRICE_RATE_OFFSET, _RATE_BIT_LENGTH);
     }
 
     /**
-     * @dev Returns the current rate of a price rate cache.
+     * @dev Returns the old rate in the price rate cache.
      */
-    function getPostJoinExitRate(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_POST_JOINEXIT_RATE_OFFSET, 96);
+    function getOldRate(bytes32 cache) internal pure returns (uint256) {
+        return cache.decodeUint(_OLD_PRICE_RATE_OFFSET, _RATE_BIT_LENGTH);
     }
 
     /**
-     * @dev Sets the post join/exit rate in an existing cache record, and returns it.
+     * @dev Copies the current rate to the old rate.
      */
-    function setPostJoinExitRate(bytes32 cache, uint256 rate) internal pure returns (bytes32) {
-        _require(rate >> 96 == 0, Errors.PRICE_RATE_OVERFLOW);
-
-        return cache.insertUint(rate, _POST_JOINEXIT_RATE_OFFSET, 96);
+    function updateOldRate(bytes32 cache) internal pure returns (bytes32) {
+        return cache.insertUint(getCurrentRate(cache), _OLD_PRICE_RATE_OFFSET, _RATE_BIT_LENGTH);
     }
 
     /**
      * @dev Returns the duration of a price rate cache.
      */
     function getDuration(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, 32);
+        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, _DURATION_BIT_LENGTH);
     }
 
     /**
      * @dev Returns the duration and expiration time of a price rate cache.
      */
     function getTimestamps(bytes32 cache) internal pure returns (uint256 duration, uint256 expires) {
         duration = getDuration(cache);
-        expires = cache.decodeUint(_PRICE_RATE_CACHE_EXPIRES_OFFSET, 32);
+        expires = cache.decodeUint(_PRICE_RATE_CACHE_EXPIRES_OFFSET, _DURATION_BIT_LENGTH);
     }
 
     /**
      * @dev Encodes rate and duration into a price rate cache. The expiration time is computed automatically, counting
      * from the current time.
      */
-    function encode(uint256 rate, uint256 duration) internal view returns (bytes32) {
-        _require(rate >> 96 == 0, Errors.PRICE_RATE_OVERFLOW);
+    function updateRate(
+        bytes32 cache,
+        uint256 rate,
+        uint256 duration
+    ) internal view returns (bytes32) {
+        _require(rate >> _RATE_BIT_LENGTH == 0, Errors.PRICE_RATE_OVERFLOW);
 
         // solhint-disable not-rely-on-time
         return
-            WordCodec.encodeUint(rate, _CURRENT_PRICE_RATE_OFFSET, 96) |
-            WordCodec.encodeUint(duration, _PRICE_RATE_CACHE_DURATION_OFFSET, 32) |
-            WordCodec.encodeUint(block.timestamp + duration, _PRICE_RATE_CACHE_EXPIRES_OFFSET, 32);
+            cache
+                .insertUint(rate, _CURRENT_PRICE_RATE_OFFSET, _RATE_BIT_LENGTH)
+                .insertUint(duration, _PRICE_RATE_CACHE_DURATION_OFFSET, _DURATION_BIT_LENGTH)
```

Comment:
Since we're now inserting, we could not update the duration here, and have a separate `setDuration` function.

Reply:
We could, but the only place it's used sets both at the same time. I renamed to setRateAndDuration, so that it's accurate (especially since we have an update for oldRate, and it could be confusing that something called `setRate` had a side effect).

I also added separate updateCurrentRate and updateDuration functions, even though they're not used. (It's in a library, so no extra bytecode.) We seem to be using `update` rather than `set` in this library, so I kept that convention.

Reply:
Both are set at the same time, but it you look at the callsites, the duration is either read right before the call, or we're updating the duration and not the rate.

e.g.:

```
function updateTokenRateCache(IERC20 token) public virtual {
    IRateProvider provider = _getRateProvider(token);
    _require(address(provider) != address(0), Errors.TOKEN_DOES_NOT_HAVE_RATE_PROVIDER);
    uint256 duration = _tokenRateCaches[token].getDuration(); // <----
    _updateTokenRateCache(token, provider, duration);
}
```

```
function setTokenRateCacheDuration(IERC20 token, uint256 duration) external authenticate {
    IRateProvider provider = _getRateProvider(token);
    _require(address(provider) != address(0), Errors.TOKEN_DOES_NOT_HAVE_RATE_PROVIDER);
    _updateTokenRateCache(token, provider, duration); // <----- here we don't care about rate
    emit TokenRateProviderSet(token, provider, duration);
}
```

Reply:
Yes, I could do this, using the new functions:

```
    function updateTokenRateCache(IERC20 token) public virtual {
        IRateProvider provider = _getRateProvider(token);
        _require(address(provider) != address(0), Errors.TOKEN_DOES_NOT_HAVE_RATE_PROVIDER);
        bytes32 cache = _tokenRateCaches[token];
        uint256 rate = provider.getRate();

        _tokenRateCaches[token] = cache.updateCurrentRate(rate);
        emit TokenRateCacheUpdated(token, rate);
    }

    function setTokenRateCacheDuration(IERC20 token, uint256 duration) external authenticate {
        IRateProvider provider = _getRateProvider(token);
        _require(address(provider) != address(0), Errors.TOKEN_DOES_NOT_HAVE_RATE_PROVIDER);
        bytes32 cache = _tokenRateCaches[token];

        _tokenRateCaches[token] = cache.updateDuration(duration);
        emit TokenRateProviderSet(token, provider, duration);
    }
```
But the tests fail then, because they expect `expires` to get updated based on the current timestamp of the update. `expires` still changes, even if duration stays the same.

The `itUpdatesTheCacheDuration` test also checks the rate, even though it supposedly only updates the duration. I'm not sure what the intent was here: are the tests wrong? *Should* it always update `expires` on every update, even if only the rate is changing?

Reply:
tbh this is a bit beside the point. Let's leave as-is for now and maybe revisit this later.